_posts/2023-02-16-coprs-migration-to-github.md

```diff
@@ -0,0 +1,145 @@
+---
+title: Copr's migration to GitHub
+author: Jiří Kyjovský
+layout: post
+---
+
+For a very long time, Copr mainliners have been considering migrating from
+Pagure to GitHub. The problem has been the constant problems with Pagure
+and even though GitHub is not opensource, it just works unlike Pagure. Not
+to mention that GitHub is miles ahead of Pagure with plenty of useful
+features.
+
+So on the migration itself, Copr had a mirror on GitHub at the time, so
+the code migration wasn't a problem. The problem was migrating all the issues
+and pull requests so that their IDs were preserved. We wanted to keep the
+same IDs for compatibility with the links in the commits. And there
+is no general solution to migrate all issues and PRs from Pagure to GitHub
+while preserving the IDs. So that's why I wrote a temporary
+[script](https://github.com/nikromen/copr-gh-migration/blob/main/script.py)
+to migrate Copr. Please note that the code was only meant to be used once
+and only for the Copr migration, so it's pretty much a piece of garbage
+and should not be used by anyone. However, it can serve as inspiration
+and I would also like to share with you some of the problems I encountered
+during the migration.
+
+
+## Choose the right tools
+
+For the whole migration I used the tool [ogr](https://github.com/packit/ogr),
+which provides a unified API for communicating with GitHub, Gitlab and Pagure.
+For communication with the GitHub API, you need to set up a
+[GitHub token](https://github.com/settings/tokens) and likewise for
+[Pagure](https://pagure.io/settings#nav-api-tab). Note the permissions that
+you give the token, only give it as much as you really need (for example, you
+probably don't want to allow it to delete repository).
+
+And then you just have to start using the tokens
+
+```py
+gh_service = GithubService(token=GITHUB_TOKEN)
+pagure_service = PagureService(token=PAGURE_TOKEN, instance_url="https://pagure.io")
+
+gh_copr_project = gh_service.get_project(namespace="fedora-copr", repo="copr")
+pagure_copr_project = pagure_service.get_project(namespace="copr", repo="copr", username="pagure_nick")
+```
+
+
+## What to migrate
+
+First of all, it is important to determine what we actually want to migrate.
+In our case we wanted to migrate as quickly as possible, so we just needed
+to migrate all issues and PRs to keep the order of their IDs. In order to
+save time, decided to migrate the PRs as issues while migrating only
+the Issues content. For PRs, we put a link to the original PR in Pagure.
```

Comment:
It would be worth mentioning that we migrated private issues by creating only dummy issues on GitHub to create a gap between IDs.